# Fix rules_rust_prost handling of transitive proto dependencies.
#
# To ensure prost_build only generates structs for messages in the proto being
# compiled, all external types must be provided as extern_paths -- not just
# types defined in direct dependencies.
#
# See https://github.com/bazelbuild/rules_rust/issues/2185.
--- providers.bzl
+++ providers.bzl
@@ -8,1 +8,1 @@
-        "package_info": "File: A newline delimited file of `--extern_path` values for protoc.",
+        "package_info": "depset[File]: Newline delimited files of `--extern_path` values for protoc.",

--- private/prost.bzl
+++ private/prost.bzl
@@ -58,1 +58,1 @@
-    dep_package_infos = [dep[ProstProtoInfo].package_info for dep in deps]
+    dep_package_infos = depset(transitive = [dep[ProstProtoInfo].package_info for dep in deps]).to_list()
@@ -109,4 +109,4 @@
     additional_inputs = depset(
-        [deps_info_file, proto_info.direct_descriptor_set] + [dep[ProstProtoInfo].package_info for dep in deps],
-        transitive = [all_additional_srcs],
+        [deps_info_file, proto_info.direct_descriptor_set],
+        transitive = [all_additional_srcs] + [dep[ProstProtoInfo].package_info for dep in deps],
     )
@@ -352,5 +352,5 @@
         ProstProtoInfo(
             dep_variant_info = dep_variant_info,
             transitive_dep_infos = depset(transitive = transitive_deps),
-            package_info = package_info_file,
+            package_info = depset([package_info_file], transitive = [dep[ProstProtoInfo].package_info for dep in proto_deps]),
         ),
